pkgs/tools/virtualization/nixos-container/default.nix

```diff
@@ -0,0 +1,22 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  nixos-container = pkgs.substituteAll {
+    name = "nixos-container";
+    dir = "bin";
+    isExecutable = true;
+    src = ./nixos-container.pl;
+    perl = "${pkgs.perl}/bin/perl -I${pkgs.perlPackages.FileSlurp}/lib/perl5/site_perl";
+    su = "${pkgs.shadow.su}/bin/su";
+    inherit (pkgs) utillinux;
+
+    postInstall = ''
+      t=$out/etc/bash_completion.d
+      mkdir -p $t
+      cp ${./nixos-container-completion.sh} $t/nixos-container
+    '';
+  };
+in nixos-container
```

Comment:
The `let nixos-container = ... in nixos-container` is redundant here, also instead of pulling in `pkgs` as a whole I'd refer to the dependencies explicitly in the expressions input attributes, like:

```nix
{ substituteAll, perl, perlPackages, shadow, utillinux }:

substituteAll {
  ...
}
```

Also, `lib` isn't used either.

Another thing that is missing is the `callPackage` from `all-packages.nix`.